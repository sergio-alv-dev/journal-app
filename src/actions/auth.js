import { types } from "../types/types";
import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { uiFinishLoading, uiStartLoading } from "./ui";
import Swal from 'sweetalert2';
import { notesLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    
    
    return ( dispatch ) => {
        
        dispatch( uiStartLoading() );

         return firebase.auth().signInWithEmailAndPassword( email, password )
            .then( ({ user }) => {
                // console.log(user);
                dispatch( login(user.uid, user.displayName ) );
                dispatch( uiFinishLoading() );
            })
            .catch( e => {
                console.log(e);
                dispatch( uiFinishLoading() );
                Swal.fire('Error', e.message, 'error');
            })
        

    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return( dispatch ) => {
        firebase.auth().createUserWithEmailAndPassword( email, password )
        .then( async ({ user }) => {
            
            await user.updateProfile({ displayName:name });
            console.log(user);
            dispatch(
                login(user.uid, user.displayName )
            )
            
        })
        .catch( e => {
            console.log(e)
            Swal.fire('Error', e.message, 'error');
        })
    }

}

export const startGoogleLogin = () => {
    return ( dispatch ) => {

        // console.log( 'google' );

        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch(
                    login(user.uid, user.displayName )
                )
                // console.log( userCred );
            })
            .catch( e => {
                console.log(e)
                Swal.fire('Error', e.message, 'error');

            })

    }
}


export const login = ( uid, displayName) => {
    return{
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const startLogout = () => {
    return async ( dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() );
        dispatch( notesLogout() );
    }
}

export const logout = () => ({
    type: types.logout
})