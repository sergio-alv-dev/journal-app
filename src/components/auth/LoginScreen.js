import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { loading } = useSelector( state => state.ui);

    const [ formValues, handleInputChange ] = useForm({
        email: '',
        password: ''
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if ( isFormValid() ){
            dispatch( startLoginEmailPassword( email, password ) );
        }
    }

    const handleGoogleLogin = () => { 
        // console.log('click handleGoogleLogin')
        dispatch ( startGoogleLogin() );
    };

    const isFormValid = () => {

        if ( !validator.isEmail( email ) ) {
            dispatch( setError('Email is not valid') );
            return false;
        } else if ( password.length < 5 ) {
            dispatch( setError('Password is not valid') );
            return false;
        }
            
        dispatch( removeError() );
        
        return true;
    }

    return (
        <>
            <div
                className = "animate__animated animate__fadeIn"    
                >

            <h3 className= "auth__title mb-5"> Login </h3>

            <form 
                onSubmit = { handleLogin }

            >
                <input 
                    type = "text"
                    placeholder = "Email"
                    name = "email"
                    className = "auth__input"
                    autoComplete = "off"
                    value = { email }
                    onChange = { handleInputChange }
                />
                
                <input 
                    type = "password"
                    placeholder = "Password"
                    name = "password"
                    className = "auth__input"
                    value = { password }
                    onChange = { handleInputChange }
                />

                <button
                    type = "submit"
                    className = "btn btn-primary btn-block"
                    disabled = { loading }
                >
                    Login
                </button>
                
                <div className = "auth__social-networks mb-1">
                    <p>Login with Social networks</p>

                    <div 
                        className="google-btn" 
                        onClick = { handleGoogleLogin }
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <hr/>
                <Link 
                    to = "/auth/register"
                    className = "link"    
                >
                Create new Account
                </Link>
            </form>
        </div>

        </>
    )
}
