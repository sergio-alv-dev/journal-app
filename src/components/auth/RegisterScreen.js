import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';

import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

// se puede utilizar useState, no es necesario utilizar redux aqui
    const dispatch = useDispatch();
    const { loading, msgError } = useSelector( state => state.ui );
    
    const [ formValues, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });
    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        
        if ( isFormValid() ) {
            dispatch( startRegisterWithEmailPasswordName( email, password, name ));
        }
    }

    const isFormValid = () => {

        if( name.trim().length === 0){
            dispatch( setError('Name is required') );
            return false;
        } else if ( !validator.isEmail( email ) ) {
            dispatch( setError('Email is not valid') );
            return false;
        } else if ( password !== password2 || password.length < 5 ) {
            dispatch( setError('Password is not valid') );
            return false;
        } else {
            dispatch( removeError() );
        }

        return true;
    }



    return (
        <>
        < div
                className = "animate__animated animate__fadeIn"    >

            
            <h3 className= "auth__title mb-5"> Register </h3>
            
            {
                msgError 
                &&
                (
                    <div className="auth__alert-error">
                        {
                            msgError 
                        }
                    </div>
                )
            }

            <form onSubmit = { handleRegister }>

                <input 
                    type = "text"
                    placeholder = "Name"
                    name = "name"
                    className = "auth__input"
                    autoComplete = "off"
                    value = { name }
                    onChange = { handleInputChange }
                />
                <input 
                    type = "text"
                    placeholder = "Email"
                    name = "email"
                    className = "auth__input"
                    autoComplete = "off"
                    value = { email }
                    onChange = { handleInputChange }
                />
                
                <input 
                    type = "password"
                    placeholder = "Password"
                    name = "password"
                    className = "auth__input"
                    value = { password }
                    onChange = { handleInputChange }
                />

                <input 
                    type = "password"
                    placeholder = "Confirm Password"
                    name = "password2"
                    className = "auth__input"
                    value = { password2 }
                    onChange = { handleInputChange }
                />

                <button
                    type = "submit"
                    className = "btn btn-primary btn-block mb-5"
                    disabled = { loading }
                >
                    Register
                </button>
                
                <hr/>
                <Link
                    to = "/auth/login"
                    className = "link mt-5"    
                >
                    Already registered?
                </Link>
            </form>
            </div>

        </>
    )
}
