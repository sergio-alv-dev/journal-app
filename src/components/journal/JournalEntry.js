import React from 'react';
import moment from 'moment';
import { activeNote } from '../../actions/notes'
import { useDispatch } from 'react-redux';

export const JournalEntry = ( { id, date, title, body, url } ) => {

    const noteDate = moment(date);
    const dispatch = useDispatch();
    
    
    const handleEntryClick = () => {
        // console.log("click entry" , id)
        dispatch( activeNote( id, {
            id, date, title, body, url
        }));

    }

    return (
        <div 
            className ="journal__entry pointer animate__animated animate__fadeIn"
            onClick = { handleEntryClick }

        >
           
           { 
            url
            &&
            (<div
                className ="journal__entry-picture"
                style = {{
                    backgroundSize: 'cover',
                    backgroundImage: `url(${ url })`
                }}
            ></div>)
           } 
            <div className ="journal__entry-body">
                <p className="journal__entry-title">
                    { title }
                </p>

                <p className="journal__entry-content">
                    { body }
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span> { noteDate.format('dddd') } </span>
                <h4> { noteDate.format('do') } </h4>
            </div>

        </div>
    )
}
