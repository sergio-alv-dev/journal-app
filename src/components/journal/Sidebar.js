import React from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {

    const dispatch = useDispatch();
    const { name } = useSelector( state => state.auth );

    const handleLogout = () => {
        dispatch( startLogout() );
    }

    const handleAddNew  = () => {
        dispatch( startNewNote() );
    }
    return (
        <aside className= "journal__side-bar">
            
            <div className ="journal_sidebar-navbar mt-1">
                <h3>
                    <i className="far fa-moon" />
                    <span> { name } </span>
                </h3>

                <button
                    className ="btn"
                    onClick = { handleLogout }
                > Logout </button>
            </div>

            <div 
                className="journal__new-entry"
                onClick = { handleAddNew }    
            >
                <i className="far fa-calendar-plus fa-5x" />
                <p className="mt-5"
                > New Entry</p>
            </div>

            <JournalEntries />


            
        </aside>
    )
}
