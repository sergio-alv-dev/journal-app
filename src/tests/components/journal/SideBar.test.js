import React from 'react';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

import { Sidebar } from "../../../components/journal/Sidebar";
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

jest.mock('../../../actions/auth', () => ({
    startLogout: jest.fn(),
}));

jest.mock('../../../actions/notes', () => ({
    startNewNote: jest.fn(),
}));


const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initState = {
    auth: {
        uid: '123',
        name: 'sergio'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{
        notes: [],
        active: {
            id: 'abc',
        }
    }
};

let store = mockStore( initState );
store.dispatch = jest.fn();

const wrapper = mount(

    <Provider store = { store }>
        <MemoryRouter>
            <Sidebar />
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en <SideBar />', () => {

    beforeEach( () => {
        store = mockStore(initState);
        jest.clearAllMocks();
    });

    test('debe de mostrarse correctamente', () => { expect( wrapper ).toMatchSnapshot() });

    test('debe de llamar el startLogout', () => {
        wrapper.find('.btn').prop('onClick')();
        expect( startLogout ).toHaveBeenCalled();
    });
    
    test('debe de llamar el startNewNote', () => {
        wrapper.find('.journal__new-entry').prop('onClick')();
        expect( startNewNote ).toHaveBeenCalled();
    });
    
    
    
});
