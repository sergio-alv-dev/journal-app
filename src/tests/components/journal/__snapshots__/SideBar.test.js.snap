// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SideBar /> debe de mostrarse correctamente 1`] = `
<aside
  className="journal__side-bar"
>
  <div
    className="journal_sidebar-navbar mt-1"
  >
    <h3>
      <i
        className="far fa-moon"
      />
      <span>
         
        sergio
         
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
       Logout 
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
       New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
